library/core/src/alloc/mod.rs

```diff
@@ -94,6 +94,18 @@ pub unsafe trait AllocRef {
     /// The returned block may have a larger size than specified by `layout.size()`, and may or may
     /// not have its contents initialized.
     ///
+    /// Note that you may not rely on this method actually getting called, even if there are calls
+    /// to it in the source. The optimizer may detect allocation/deallocation pairs that it can
+    /// instead move to stack allocations/deallocations and thus never invoke the allocator here.
```

Comment:
There doesn't even have to be a deallocation -- the allocator may also detect that the memory returned from an allocation function is unused, and remove the alloc call.

The optimizer might also be allowed to *reorder* allocation operations.

It is tempting to just say that `alloc` may not have side-effects (i.e., side-effects are some form of UB), but of course that makes no sense as the allocator will have some internal state.

Reply:
You could say that it may not have visible side-effects beyond requesting memory from the OS and freeing it again.

Reply:
First of all, defining "visible" here is really hard. Secondly, this would make things like https://docs.rs/mockalloc/0.1.0/mockalloc/ UB. And there is no reason to make them UB in my opinion -- sure, the count they will report can change depending on optimizations, but the count will still accurately reflect what happens in the given execution.

Reply:
Even beyond something like mockalloc, "real" production allocators are not side effect free anyway. jemalloc tracks allocation statistics and can do heap profiling as one example.

Reply:
Is it currently possible to disable this optimisation using some compiler option/preprocessor define?

The ability to do (de)allocation counts and sizes is widely used in products such as game engines. Especially if you're working in codebases that are not 100% safe (e.g. a Rust to C/C++ interface), it'd be important for Rust to still reliably support that behaviour - even if it'd only be supported in and if-cfg.

Reply:
> Is it currently possible to disable this optimisation using some compiler option/preprocessor define?

That is a good question to which I do not know the answer. LLVM would have to expose this as an option.

> The ability to do (de)allocation counts and sizes is widely used in products such as game engines. Especially if you're working in codebases that are not 100% safe (e.g. a Rust to C/C++ interface), it'd be important for Rust to still reliably support that behaviour - even if it'd only be supported in and if-cfg.

Note that Rust supports this at least as well as C and C++ -- at least clang will perform the kind of optimizations this is about.

Also, Allocation counts are still *supported*. What this PR enables is, basically, allocation counts being different between debug and release builds. Usually compiler have to ensure optimizations do not change program behavior; if the program prints the number of allocations to stdout, that means no allocation may be optimized away!